Client/SafeBrowsing/GSafeBrowsingCache.swift

```diff
@@ -0,0 +1,123 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+
+private struct ThreatCache: Hashable {
+    var time: TimeInterval
+    let hash: String
+    let threatType: ThreatType
+}
+
+enum SBCacheHitType {
+    case miss
+    case positive
+    case negative
+}
+
+struct SBCacheResult {
+    let threats: [ThreatType]
+    let cacheResult: SBCacheHitType
+}
+
+class SafeBrowsingCache {
+    private let cacheLock = NSRecursiveLock()
+    private var positiveCache = [String: [ThreatCache]]()
+    private var negativeCache = [String: TimeInterval]()
+    
+    func find(_ hash: String) -> SBCacheResult {
+        cacheLock.lock(); defer { cacheLock.unlock() }
+        
+        if let threats = positiveCache[hash] {
+            var results = [ThreatType]()
+            
+            for threat in threats {
+                if Date().timeIntervalSince1970 > threat.time  {
```

Comment:
Is this check reversed or am I misreading this? I see `if now > expiry`, which is true when the entry has expired.

Reply:
It has been fixed yeah. I tend to write out all the code first and then test it after. After testing, it was actually reversed and I fixed it and verified the caching works :)